src/coreclr/gc/gcpriv.h

```diff
@@ -3432,6 +3432,39 @@ class gc_heap
     PER_HEAP
     BOOL heap_analyze_success;
 
+    PER_HEAP
+    uint32_t* card_table;
+
+// For debugger access purpose. these fields are currently included unconditionally
+// in the gc_heap class. But they should not be used under certain conditions.
+// We can turn on COMPILER_CHECK_FIELDS_ACCESS below (and the corresponding one)
+// in gcinterface.dac.h and have the compiler to verify for us that it is indeed the
+// case. However, COMPILER_CHECK_FIELDS_ACCESS is incompatible with debugging and
+// must be turned off.
+
+// #define COMPILER_CHECK_FIELDS_ACCESS
+
+#if !defined(COMPILER_CHECK_FIELDS_ACCESS) || defined(BACKGROUND_GC)
```

Comment:
we should not have a field that's related to a feature (like `BACKGROUND_GC`) present when that feature is not defined in `gc_heap`. so if `BACKGROUND_GC` isn't defined, `mark_array` should not be present on the `gc_heap` class. 

I don't think we should keep these `ASSERT_OFFSETS_MATCH` checks anymore - dac_gc_heap serves a different purpose from gc_heap and we shouldn't compromise the invariant in `gc_heap` just so that the offsets for these fields can be checked. I propose we get rid of these `ASSERT_OFFSETS_MATCH` checks altogether 'cause it doesn't seem to be useful in `PopulateDacVars` we should make sure for every check of #ifdef X, there's an #else block that initializes the fields that aren't present in `gc_heap` to 0.

Reply:
It is not the `ASSERT_OFFSETS_MATCH` that is forcing us to keep the layout between `gc_heap` and `dac_gc_heap` matches. It is the fact that the debugger is using the fact that the layout of `dac_gc_heap` and `gc_heap` matches to obtain info inside the `gc_heap` object.

At the end of `gc.cpp` we did this:

```
gcDacVars->g_heaps = reinterpret_cast<dac_gc_heap***>(&gc_heap::g_heaps);
```

On the debugger side, it will use the DAC magic to dereference these pointers and get to the data we wanted. Therefore it is important to keep the layout the same if we wanted to use this existing mechanism.

Reply:
ahh I see. it's really not great to have this kind of artificially coupling between the 2 sides. 

we could have the GC side populate a dac_gc_heap array and give that to the DAC side.

Reply:
At the end of the day, the debugger needs to be able to access the values without the debuggee running (because it could be dead (e.g. a dump)).

At its core, here is the main problem:

1) `clrgc.dll` and `coreclr.dll` have different layouts for `gc_heap`, but
2) `mscordaccore.dll` hardcoded the field offsets (by compiling the definition of `dac_gc_heap` and assume `gc_heap` shared the same).

My earlier solution was simply sidestepping that core issue by artificially including unused fields. By doing so, we have an identical layout, and therefore we can continue to use the same mechanism.

Looking forward, if we going to continue to evolve the GC in this fashion, this is not going to sustain.

By avoiding the hard code, we can solve this problem. First, stop assuming that `dac_gc_heap` and `gc_heap` match, but instead, we can have the GC filling an array with the field offsets and publish that to the DAC, then the DAC can use that array to access field information. This will solve the issue cleanly.

Note that we have used a similar pattern for other structures in the GC as well, basically all the structures in `gcinterface.dac.h` have layout assumptions, since they are always the same (for now), there is no particular urgency to change that for .NET 6.